libs/chain-signatures/contract/src/lib.rs

```diff
@@ -364,14 +437,21 @@ impl VersionedMpcContract {
             {
                 return Err(RespondError::InvalidSignature.into());
             }
-
-            match self {
-                Self::V0(mpc_contract) => {
-                    handle_signature_response(mpc_contract, &request, &response)
-                }
-                Self::V1(mpc_contract) => {
-                    handle_signature_response(mpc_contract, &request, &response)
+            // First get the yield promise of the (potentially timed out) request.
+            if let Some(YieldIndex { data_id }) = self.get_pending_request(&request) {
+                // Then resolve the promise. This will have no effect if the request already timed
+                // out by the blockchain.
+                env::promise_yield_resume(&data_id, &serde_json::to_vec(&response).unwrap());
+                // Only then clean up the state.
+                // This order of execution ensures that the state is cleaned of the current
+                // response, even if it belongs to an already timed out signature request.
+                match self {
+                    Self::V0(_) | Self::V1(_) => {}
+                    Self::V2(mpc_contract) => mpc_contract.remove_timed_out_requests(),
```

Comment:
I think we may need to be careful about gas cost predictability here. If enough timeout requests accumulate, any respond call made afterwards would attempt to clear all of them and run out of gas. If that happens the transaction as a whole fails and none of the requests are cleared, causing the system to get stuck. At the same time we don't want to over-attach gas to respond calls since it limits throughput.

One idea to prevent this would be to limit the number of timed out requests removed per call to some reasonable number.

Reply:
Also maybe it would be nicer to trigger the cleanup from "sign" calls rather than "respond" calls? Then we don't have the possibility that state accumulates without limit if the signer network happens to be down. Of course ideally that won't ever happen...

Reply:
cleanup is now triggered in `sign()` and `respond()`. I also added a function that can be called to trigger the cleanup separately.